File: workers/classifier/src/schemas.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any

SUPPORTED_PRODUCT_PROFILES = {
    "adc_dac_converter",
    "rf_transceiver",
    "mcu_processor_soc",
    "fpga_programmable_logic_soc",
    "crypto_security_device",
    "generic_electronics",
}

CONFIDENCE_LEVELS = {"high", "medium", "low"}


@dataclass
class WorkerInput:
    document_id: str
    document_title: str
    file_path: str
    organization_id: str
    document_metadata: dict[str, Any]


@dataclass
class ExtractedSpec:
    name: str
    value: str
    unit: str | None
    source_snippet: str
    importance: str
    category: str
    confidence: str
    display_name: str | None = None
    source_text: str | None = None
    source_page_from: int | None = None
    source_page_to: int | None = None
    bounding_boxes: list[dict[str, float]] | None = None
    extraction_rationale: str = "Captured from source-backed technical text."
    value_type: str = "directly_stated"
    extraction_method: str = "python_worker"
    extraction_method_version: str = "v4"


@dataclass
class RegulationSource:
    authority: str
    regulation_title: str
    citation_text: str
    kind: str
    verification_status: str
    regulation_version: str | None = None
    citation_url: str | None = None
    source_identifier: str | None = None
    section: str | None = None
    paragraph: str | None = None
    last_verified_at: str | None = None


@dataclass
class RegulatoryCitation:
    citation_label: str
    citation_text: str
    source: str
    relevance: str
    regulation_source: RegulationSource | None = None


@dataclass
class FactIssue:
    issue_type: str
    summary: str
    details: str | None = None
    primary_fact_name: str | None = None
    related_fact_name: str | None = None


@dataclass
class CandidateFactMapping:
    fact_name: str
    criterion_label: str
    matched_value: str
    comparison_result: str
    notes: str | None = None


@dataclass
class ReviewPath:
    path_key: str
    title: str
    scope: str
    type: str
    status: str
    why_triggered: str
    triggered_fact_names: list[str]
    regulatory_citations: list[RegulatoryCitation]
    missing_information: list[str]
    reviewer_questions: list[str]
    technical_risk_area: str | None = None
    reviewer_notes: str | None = None
    decision_rationale: str | None = None


@dataclass
class ECCNCandidate:
    eccn: str
    title: str
    confidence: str
    matched_technical_facts: list[str]
    regulatory_citations: list[RegulatoryCitation]
    why_it_may_apply: str
    why_it_may_not_apply: str
    missing_information: list[str]
    uncertainty_flags: list[str]
    reviewer_questions: list[str]
    official_title: str | None = None
    confidence_rationale: str = "Evidence completeness remains limited."
    status: str = "review_required"
    regulation_source: RegulationSource | None = None
    paragraph_reference: str | None = None
    control_criteria: list[str] | None = None
    fact_mappings: list[CandidateFactMapping] | None = None
    may_apply_reasons: list[str] | None = None
    may_not_apply_reasons: list[str] | None = None
    alternative_candidates: list[dict[str, str]] | None = None
    review_path_id: str | None = None
    review_path_key: str | None = None


@dataclass
class WorkerOutput:
    document_id: str
    organization_id: str
    requires_human_review: bool
    confidence: float
    confidence_rationale: str
    uncertainty_flags: list[str]
    extracted_specs: list[ExtractedSpec]
    fact_issues: list[FactIssue]
    review_paths: list[ReviewPath]
    eccn_candidates: list[ECCNCandidate]
    memo_markdown: str
    artifacts: dict[str, str]
    run_metadata: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass
class ProductProfile:
    profile: str
    confidence: str
    rationale: str
    supporting_snippets: list[str]
    secondary_profiles: list[str]


@dataclass
class ProductIdentity:
    manufacturer: str | None
    product_name: str | None
    product_family: str | None
    part_number: str | None
    document_number: str | None
    document_type: str | None
    is_family_overview: bool


@dataclass
class MissingFact:
    name: str
    category: str
    why_it_matters: str
    status: str


@dataclass
class AIExtractionResult:
    product_profile: ProductProfile
    product_identity: ProductIdentity
    extracted_facts: list[ExtractedSpec]
    missing_facts: list[MissingFact]
    warnings: list[str]
    raw: dict[str, Any]


def _require_dict(value: Any, path: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(f"{path} must be an object")
    return value


def _require_str(value: Any, path: str, *, allow_empty: bool = False) -> str:
    if not isinstance(value, str):
        raise ValueError(f"{path} must be a string")
    clean = value.strip()
    if not allow_empty and not clean:
        raise ValueError(f"{path} must not be empty")
    return clean


def _optional_str(value: Any, path: str) -> str | None:
    if value is None:
        return None
    clean = _require_str(value, path, allow_empty=True)
    return clean or None


def _require_bool(value: Any, path: str) -> bool:
    if not isinstance(value, bool):
        raise ValueError(f"{path} must be a boolean")
    return value


def _string_list(value: Any, path: str) -> list[str]:
    if value is None:
        return []
    if not isinstance(value, list):
        raise ValueError(f"{path} must be a list")
    return [_require_str(item, f"{path}[{index}]") for index, item in enumerate(value) if str(item).strip()]


def _confidence(value: Any, path: str) -> str:
    clean = _require_str(value, path).lower()
    if clean not in CONFIDENCE_LEVELS:
        raise ValueError(f"{path} must be one of high, medium, low")
    return clean


def _profile(value: Any, path: str) -> str:
    clean = _require_str(value, path)
    if clean not in SUPPORTED_PRODUCT_PROFILES:
        raise ValueError(f"{path} must be one of {', '.join(sorted(SUPPORTED_PRODUCT_PROFILES))}")
    return clean


def _dedupe_facts(facts: list[ExtractedSpec]) -> list[ExtractedSpec]:
    deduped: list[ExtractedSpec] = []
    seen: set[tuple[str, str, str]] = set()
    for fact in facts:
        key = (fact.name.lower(), fact.value.lower(), fact.source_snippet.lower())
        if key in seen:
            continue
        seen.add(key)
        deduped.append(fact)
    return deduped


def validate_ai_extraction_payload(payload: dict[str, Any]) -> AIExtractionResult:
    root = _require_dict(payload, "extraction")
    profile_payload = _require_dict(root.get("productProfile"), "productProfile")
    identity_payload = _require_dict(root.get("productIdentity"), "productIdentity")

    product_profile = ProductProfile(
        profile=_profile(profile_payload.get("profile"), "productProfile.profile"),
        confidence=_confidence(profile_payload.get("confidence"), "productProfile.confidence"),
        rationale=_require_str(profile_payload.get("rationale"), "productProfile.rationale"),
        supporting_snippets=_string_list(profile_payload.get("supportingSnippets"), "productProfile.supportingSnippets"),
        secondary_profiles=[
            secondary
            for secondary in _string_list(profile_payload.get("secondaryProfiles"), "productProfile.secondaryProfiles")
            if secondary in SUPPORTED_PRODUCT_PROFILES
        ],
    )

    product_identity = ProductIdentity(
        manufacturer=_optional_str(identity_payload.get("manufacturer"), "productIdentity.manufacturer"),
        product_name=_optional_str(identity_payload.get("productName"), "productIdentity.productName"),
        product_family=_optional_str(identity_payload.get("productFamily"), "productIdentity.productFamily"),
        part_number=_optional_str(identity_payload.get("partNumber"), "productIdentity.partNumber"),
        document_number=_optional_str(identity_payload.get("documentNumber"), "productIdentity.documentNumber"),
        document_type=_optional_str(identity_payload.get("documentType"), "productIdentity.documentType"),
        is_family_overview=_require_bool(identity_payload.get("isFamilyOverview"), "productIdentity.isFamilyOverview"),
    )

    facts_payload = root.get("extractedFacts")
    if not isinstance(facts_payload, list):
        raise ValueError("extractedFacts must be a list")

    facts: list[ExtractedSpec] = []
    for index, item in enumerate(facts_payload):
        fact = _require_dict(item, f"extractedFacts[{index}]")
        facts.append(
            ExtractedSpec(
                name=_require_str(fact.get("name"), f"extractedFacts[{index}].name"),
                display_name=_require_str(fact.get("displayName"), f"extractedFacts[{index}].displayName"),
                value=_require_str(fact.get("value"), f"extractedFacts[{index}].value"),
                unit=_optional_str(fact.get("unit"), f"extractedFacts[{index}].unit"),
                category=_require_str(fact.get("category"), f"extractedFacts[{index}].category"),
                source_snippet=_require_str(fact.get("sourceSnippet"), f"extractedFacts[{index}].sourceSnippet"),
                importance=_require_str(fact.get("importance"), f"extractedFacts[{index}].importance"),
                confidence=_confidence(fact.get("confidence"), f"extractedFacts[{index}].confidence"),
            )
        )

    missing_payload = root.get("missingFacts") or []
    if not isinstance(missing_payload, list):
        raise ValueError("missingFacts must be a list")
    missing_facts = [
        MissingFact(
            name=_require_str(_require_dict(item, f"missingFacts[{index}]").get("name"), f"missingFacts[{index}].name"),
            category=_require_str(_require_dict(item, f"missingFacts[{index}]").get("category"), f"missingFacts[{index}].category"),
            why_it_matters=_require_str(_require_dict(item, f"missingFacts[{index}]").get("whyItMatters"), f"missingFacts[{index}].whyItMatters"),
            status=_require_str(_require_dict(item, f"missingFacts[{index}]").get("status"), f"missingFacts[{index}].status"),
        )
        for index, item in enumerate(missing_payload[:12])
    ]

    return AIExtractionResult(
        product_profile=product_profile,
        product_identity=product_identity,
        extracted_facts=_dedupe_facts(facts),
        missing_facts=missing_facts,
        warnings=_string_list(root.get("warnings"), "warnings"),
        raw=root,
    )

File: packages/shared/src/schemas.ts
import { z } from 'zod';

export const confidenceLevelSchema = z.enum(['high', 'medium', 'low']);
export const membershipRoleSchema = z.enum([
  'OWNER',
  'ADMIN',
  'REVIEWER',
  'ANALYST',
  'VIEWER',
]);

export const uncertaintyFlagSchema = z.enum([
  'missing_key_specs',
  'ambiguous_datasheet_language',
  'multiple_plausible_eccns',
  'limited_regulatory_coverage',
  'requires_engineering_confirmation',
  'family_level_source_requires_device_specific_confirmation',
  'crypto_relevance_requires_qualified_review',
  'conflicting_source_values',
  'missing_regulatory_mapping',
]);

export const reviewWorkflowStateSchema = z.enum([
  'draft_generated',
  'awaiting_reviewer_assignment',
  'in_technical_review',
  'needs_additional_documentation',
  'escalated',
  'reviewer_conclusion_recorded',
  'approved_for_internal_use',
  'closed',
]);

export const reviewPathStatusSchema = z.enum([
  'open',
  'excluded_by_reviewer',
  'needs_more_evidence',
  'escalated',
  'resolved',
]);

export const reviewPathTypeSchema = z.enum([
  'product_area',
  'technical_risk',
  'encryption_security',
  'special_environment',
  'military_space',
  'general_fallback',
]);

export const factValueTypeSchema = z.enum([
  'directly_stated',
  'inferred',
  'normalized',
  'calculated',
]);

export const factVerificationStatusSchema = z.enum([
  'unreviewed',
  'verified',
  'corrected',
  'rejected',
  'suppressed',
]);

export const factIssueTypeSchema = z.enum([
  'contradiction',
  'duplicate',
  'family_scope_warning',
  'unverified_identifier',
  'ambiguous_unit',
]);

export const regulationSourceKindSchema = z.enum([
  'primary_regulation',
  'agency_guidance',
  'internal_playbook',
  'reviewer_note',
]);

export const regulationVerificationStatusSchema = z.enum([
  'current',
  'needs_verification',
  'archived',
  'superseded',
]);

export const eccnCandidateStatusSchema = z.enum([
  'proposed',
  'approved',
  'rejected',
  'modified',
  'review_required',
]);

export const reviewSubmissionStatusSchema = z.enum([
  'pending_review',
  'reviewed',
  'needs_more_information',
  'approved',
  'rejected',
]);

export const documentCreateSchema = z.object({
  title: z.string().min(1).max(255),
  fileName: z.string().min(1).max(255),
  mimeType: z.string().min(1).max(120),
  storagePath: z.string().min(1),
  sizeBytes: z.number().int().nonnegative().optional(),
  rawText: z.string().optional(),
  sourceType: z.enum(['upload', 'seed', 'manual']).default('upload'),
  documentType: z.string().trim().max(120).optional().or(z.literal('')),
  manufacturer: z.string().trim().max(255).optional().or(z.literal('')),
  sourceUrl: z.string().trim().url().max(2000).optional().or(z.literal('')),
  versionLabel: z.string().trim().max(120).optional().or(z.literal('')),
  sha256: z.string().trim().length(64).optional().or(z.literal('')),
  pageCount: z.number().int().positive().optional(),
  extractionStatus: z.enum(['pending', 'completed', 'failed']).default('completed'),
  origin: z.enum(['public', 'customer_provided', 'internal']).default('customer_provided'),
  visibility: z.enum(['private', 'organization', 'public_demo']).default('private'),
});

export const classificationRunCreateSchema = z.object({
  trigger: z.enum(['manual', 'api', 'reprocess']).default('manual'),
});

export const reviewSubmissionSchema = z.object({
  status: reviewSubmissionStatusSchema,
  workflowState: reviewWorkflowStateSchema.default('in_technical_review'),
  note: z.string().trim().max(4000).optional().default(''),
  approvalScope: z.string().trim().max(255).optional().or(z.literal('')),
  finalInternalRecommendation: z.string().trim().max(4000).optional().or(z.literal('')),
  caveats: z.string().trim().max(4000).optional().or(z.literal('')),
  assumptions: z.string().trim().max(4000).optional().or(z.literal('')),
  missingInformation: z.string().trim().max(4000).optional().or(z.literal('')),
});

export const publicDemoPublishSchema = z.object({
  confirmation: z.literal(true),
  publicTitle: z.string().trim().max(160).optional().or(z.literal('')),
  publicSummary: z.string().trim().max(600).optional().or(z.literal('')),
  sourceDocumentDisplayName: z
    .string()
    .trim()
    .max(255)
    .optional()
    .or(z.literal('')),
});

export const factReviewUpdateSchema = z.object({
  reviewerStatus: factVerificationStatusSchema,
  reviewerNote: z.string().trim().max(4000).optional().or(z.literal('')),
  reviewerCorrectedValue: z.string().trim().max(1000).optional().or(z.literal('')),
  reviewerCorrectedUnit: z.string().trim().max(120).optional().or(z.literal('')),
  suppressFromMemo: z.boolean().optional().default(false),
});

export const reviewPathUpdateSchema = z.object({
  status: reviewPathStatusSchema,
  reviewerNotes: z.string().trim().max(4000).optional().or(z.literal('')),
  decisionRationale: z.string().trim().max(4000).optional().or(z.literal('')),
  missingInformation: z.array(z.string().trim().min(1).max(500)).max(20).optional().default([]),
  reviewerQuestions: z.array(z.string().trim().min(1).max(500)).max(20).optional().default([]),
});

export const eccnCandidateUpdateSchema = z.object({
  status: eccnCandidateStatusSchema,
  reviewerDisposition: z.string().trim().max(255).optional().or(z.literal('')),
  reviewerDispositionRationale: z.string().trim().max(4000).optional().or(z.literal('')),
  confidenceRationale: z.string().trim().max(4000).optional().or(z.literal('')),
  alternativeCandidates: z
    .array(
      z.object({
        eccn: z.string().trim().min(1).max(32),
        title: z.string().trim().min(1).max(255),
        status: z.enum(['considered', 'excluded', 'open']),
        rationale: z.string().trim().min(1).max(1000),
      }),
    )
    .max(10)
    .optional()
    .default([]),
});

const passwordSchema = z.string().min(12).max(256);

export const signUpSchema = z
  .object({
    name: z.string().trim().min(1).max(120),
    email: z.string().trim().email().max(320),
    password: passwordSchema,
    confirmPassword: z.string().max(256),
    inviteToken: z.string().trim().optional(),
  })
  .superRefine((value, ctx) => {
    if (value.password !== value.confirmPassword) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['confirmPassword'],
        message: 'Passwords do not match.',
      });
    }
  });

export const signInSchema = z.object({
  email: z.string().trim().email().max(320),
  password: z.string().max(256),
  inviteToken: z.string().trim().optional(),
});

export const verifyEmailSchema = z.object({
  token: z.string().trim().min(1).max(512),
});

export const resendVerificationSchema = z.object({
  email: z.string().trim().email().max(320),
});

export const forgotPasswordSchema = z.object({
  email: z.string().trim().email().max(320),
});

export const resetPasswordSchema = z
  .object({
    token: z.string().trim().min(1).max(512),
    password: passwordSchema,
    confirmPassword: z.string().max(256),
  })
  .superRefine((value, ctx) => {
    if (value.password !== value.confirmPassword) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['confirmPassword'],
        message: 'Passwords do not match.',
      });
    }
  });

export const changePasswordSchema = z
  .object({
    currentPassword: z.string().max(256),
    newPassword: passwordSchema,
    confirmPassword: z.string().max(256),
  })
  .superRefine((value, ctx) => {
    if (value.newPassword !== value.confirmPassword) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['confirmPassword'],
        message: 'Passwords do not match.',
      });
    }
  });

export const onboardingSchema = z.object({
  organizationName: z.string().trim().min(1).max(120),
  industry: z.string().trim().max(120).optional().or(z.literal('')),
});

export const organizationUpdateSchema = z.object({
  name: z.string().trim().min(1).max(120),
  industry: z.string().trim().max(120).optional().or(z.literal('')),
});

export const inviteCreateSchema = z.object({
  email: z.string().trim().email().max(320),
  role: membershipRoleSchema.default('REVIEWER'),
});

export const inviteAcceptSchema = z.object({
  token: z.string().trim().min(1).max(512),
});

export const profileUpdateSchema = z.object({
  name: z.string().trim().min(1).max(120),
});

export const regulatorySourceSchema = z.object({
  authority: z.string().trim().min(1).max(255),
  regulationTitle: z.string().trim().min(1).max(255),
  regulationVersion: z.string().trim().max(120).optional().nullable(),
  citationText: z.string().trim().min(1).max(4000),
  citationUrl: z.string().trim().url().max(2000).optional().nullable(),
  sourceIdentifier: z.string().trim().max(255).optional().nullable(),
  section: z.string().trim().max(255).optional().nullable(),
  paragraph: z.string().trim().max(255).optional().nullable(),
  kind: regulationSourceKindSchema,
  lastVerifiedAt: z.string().datetime().optional().nullable(),
  verificationStatus: regulationVerificationStatusSchema,
});

export const citationSchema = z.object({
  citationLabel: z.string().trim().min(1).max(255),
  citationText: z.string().trim().min(1).max(4000),
  source: z.string().trim().min(1).max(500),
  relevance: z.string().trim().min(1).max(2000),
  regulationSource: regulatorySourceSchema,
});

export const extractedFactSchema = z.object({
  name: z.string().trim().min(1).max(255),
  displayName: z.string().trim().min(1).max(255),
  value: z.string().trim().min(1).max(1000),
  unit: z.string().trim().max(120).optional().nullable(),
  category: z.string().trim().min(1).max(255),
  sourceSnippet: z.string().trim().min(1).max(4000),
  sourceText: z.string().trim().min(1).max(4000).optional().nullable(),
  sourcePageFrom: z.number().int().nonnegative().optional().nullable(),
  sourcePageTo: z.number().int().nonnegative().optional().nullable(),
  boundingBoxes: z.array(z.record(z.string(), z.number())).optional().nullable(),
  importance: z.string().trim().min(1).max(2000),
  confidence: confidenceLevelSchema,
  extractionRationale: z.string().trim().min(1).max(2000),
  valueType: factValueTypeSchema,
  extractionMethod: z.string().trim().min(1).max(255),
  extractionMethodVersion: z.string().trim().min(1).max(255),
});

export const factIssueSchema = z.object({
  issueType: factIssueTypeSchema,
  summary: z.string().trim().min(1).max(500),
  details: z.string().trim().max(4000).optional().nullable(),
  primaryFactName: z.string().trim().max(255).optional().nullable(),
  relatedFactName: z.string().trim().max(255).optional().nullable(),
});

export const reviewPathSchema = z.object({
  pathKey: z.string().trim().min(1).max(255),
  title: z.string().trim().min(1).max(255),
  scope: z.string().trim().min(1).max(1000),
  type: reviewPathTypeSchema,
  status: reviewPathStatusSchema.default('open'),
  whyTriggered: z.string().trim().min(1).max(2000),
  technicalRiskArea: z.string().trim().max(255).optional().nullable(),
  triggeredFactNames: z.array(z.string().trim().min(1).max(255)).max(30),
  regulatoryCitations: z.array(citationSchema),
  missingInformation: z.array(z.string().trim().min(1).max(500)).max(20),
  reviewerQuestions: z.array(z.string().trim().min(1).max(500)).max(20),
  reviewerNotes: z.string().trim().max(4000).optional().nullable(),
  decisionRationale: z.string().trim().max(4000).optional().nullable(),
});

export const candidateFactMappingSchema = z.object({
  factName: z.string().trim().min(1).max(255),
  criterionLabel: z.string().trim().min(1).max(255),
  matchedValue: z.string().trim().min(1).max(1000),
  comparisonResult: z.string().trim().min(1).max(1000),
  notes: z.string().trim().max(2000).optional().nullable(),
});

export const alternativeCandidateSchema = z.object({
  eccn: z.string().trim().min(1).max(32),
  title: z.string().trim().min(1).max(255),
  status: z.enum(['considered', 'excluded', 'open']),
  rationale: z.string().trim().min(1).max(1000),
});

export const eccnCandidateSchema = z.object({
  eccn: z
    .string()
    .trim()
    .regex(/^[0-9][A-Z][0-9]{3}[a-zA-Z0-9]*$/)
    .max(32),
  title: z.string().trim().min(1).max(255),
  officialTitle: z.string().trim().min(1).max(255),
  confidence: confidenceLevelSchema,
  confidenceRationale: z.string().trim().min(1).max(2000),
  status: eccnCandidateStatusSchema.default('review_required'),
  regulationSource: regulatorySourceSchema.refine(
    (value) => value.kind === 'primary_regulation' || value.kind === 'agency_guidance',
    'ECCN candidates require an official regulation source.',
  ),
  paragraphReference: z.string().trim().max(255).optional().nullable(),
  controlCriteria: z.array(z.string().trim().min(1).max(1000)).max(20),
  factMappings: z.array(candidateFactMappingSchema).min(1).max(30),
  matchedTechnicalFacts: z.array(z.string().trim().min(1).max(1000)).max(30),
  whyItMayApply: z.string().trim().min(1).max(2000),
  whyItMayNotApply: z.string().trim().min(1).max(2000),
  mayApplyReasons: z.array(z.string().trim().min(1).max(1000)).max(20),
  mayNotApplyReasons: z.array(z.string().trim().min(1).max(1000)).max(20),
  missingInformation: z.array(z.string().trim().min(1).max(500)).max(20),
  uncertaintyFlags: z.array(uncertaintyFlagSchema).max(20),
  reviewerQuestions: z.array(z.string().trim().min(1).max(500)).max(20),
  alternativeCandidates: z.array(alternativeCandidateSchema).max(10),
  reviewPathKey: z.string().trim().max(255).optional().nullable(),
});

export const workerOutputSchema = z
  .object({
    documentId: z.string(),
    organizationId: z.string(),
    requiresHumanReview: z.literal(true),
    confidence: z.number().min(0).max(1),
    confidenceRationale: z.string().trim().min(1).max(2000),
    uncertaintyFlags: z.array(uncertaintyFlagSchema),
    extractedSpecs: z.array(extractedFactSchema),
    factIssues: z.array(factIssueSchema).default([]),
    reviewPaths: z.array(reviewPathSchema).default([]),
    eccnCandidates: z.array(eccnCandidateSchema).default([]),
    memoMarkdown: z.string(),
    artifacts: z.object({
      extractedTextPath: z.string(),
      structuredOutputPath: z.string(),
      memoPath: z.string(),
    }),
    runMetadata: z.record(z.string(), z.unknown()).optional().nullable(),
  })
  .superRefine((value, ctx) => {
    const hasCryptoFacts = value.extractedSpecs.some(
      (fact) =>
        fact.category === 'security_cryptography' ||
        /crypto|cryptograph|encrypt|secure boot|key management|secure element/i.test(
          `${fact.name} ${fact.displayName} ${fact.value} ${fact.sourceSnippet}`,
        ),
    );

    if (hasCryptoFacts) {
      const wronglyDenied = value.reviewPaths.some(
        (path) =>
          /crypt/i.test(path.title) &&
          /not found/i.test(path.whyTriggered),
      );
      if (wronglyDenied) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message:
            'If cryptographic capabilities are extracted, review paths cannot say crypto features were not found.',
          path: ['reviewPaths'],
        });
      }
    }
  });

export const workerCliOutputSchema = z.object({
  document_id: z.string(),
  organization_id: z.string(),
  requires_human_review: z.literal(true),
  confidence: z.number().min(0).max(1),
  confidence_rationale: z.string(),
  uncertainty_flags: z.array(uncertaintyFlagSchema),
  extracted_specs: z.array(
    z.object({
      name: z.string(),
      display_name: z.string(),
      value: z.string(),
      unit: z.string().nullable().optional(),
      category: z.string(),
      source_snippet: z.string(),
      source_text: z.string().nullable().optional(),
      source_page_from: z.number().int().nonnegative().nullable().optional(),
      source_page_to: z.number().int().nonnegative().nullable().optional(),
      bounding_boxes: z.array(z.record(z.string(), z.number())).nullable().optional(),
      importance: z.string(),
      confidence: confidenceLevelSchema,
      extraction_rationale: z.string(),
      value_type: factValueTypeSchema,
      extraction_method: z.string(),
      extraction_method_version: z.string(),
    }),
  ),
  fact_issues: z.array(
    z.object({
      issue_type: factIssueTypeSchema,
      summary: z.string(),
      details: z.string().nullable().optional(),
      primary_fact_name: z.string().nullable().optional(),
      related_fact_name: z.string().nullable().optional(),
    }),
  ),
  review_paths: z.array(
    z.object({
      path_key: z.string(),
      title: z.string(),
      scope: z.string(),
      type: reviewPathTypeSchema,
      status: reviewPathStatusSchema,
      why_triggered: z.string(),
      technical_risk_area: z.string().nullable().optional(),
      triggered_fact_names: z.array(z.string()),
      regulatory_citations: z.array(
        z.object({
          citation_label: z.string(),
          citation_text: z.string(),
          source: z.string(),
          relevance: z.string(),
          regulation_source: z.object({
            authority: z.string(),
            regulation_title: z.string(),
            regulation_version: z.string().nullable().optional(),
            citation_text: z.string(),
            citation_url: z.string().nullable().optional(),
            source_identifier: z.string().nullable().optional(),
            section: z.string().nullable().optional(),
            paragraph: z.string().nullable().optional(),
            kind: regulationSourceKindSchema,
            last_verified_at: z.string().nullable().optional(),
            verification_status: regulationVerificationStatusSchema,
          }),
        }),
      ),
      missing_information: z.array(z.string()),
      reviewer_questions: z.array(z.string()),
      reviewer_notes: z.string().nullable().optional(),
      decision_rationale: z.string().nullable().optional(),
    }),
  ),
  eccn_candidates: z.array(
    z.object({
      eccn: z.string(),
      title: z.string(),
      official_title: z.string(),
      confidence: confidenceLevelSchema,
      confidence_rationale: z.string(),
      status: eccnCandidateStatusSchema,
      regulation_source: z.object({
        authority: z.string(),
        regulation_title: z.string(),
        regulation_version: z.string().nullable().optional(),
        citation_text: z.string(),
        citation_url: z.string().nullable().optional(),
        source_identifier: z.string().nullable().optional(),
        section: z.string().nullable().optional(),
        paragraph: z.string().nullable().optional(),
        kind: regulationSourceKindSchema,
        last_verified_at: z.string().nullable().optional(),
        verification_status: regulationVerificationStatusSchema,
      }),
      paragraph_reference: z.string().nullable().optional(),
      control_criteria: z.array(z.string()),
      fact_mappings: z.array(
        z.object({
          fact_name: z.string(),
          criterion_label: z.string(),
          matched_value: z.string(),
          comparison_result: z.string(),
          notes: z.string().nullable().optional(),
        }),
      ),
      matched_technical_facts: z.array(z.string()),
      why_it_may_apply: z.string(),
      why_it_may_not_apply: z.string(),
      may_apply_reasons: z.array(z.string()),
      may_not_apply_reasons: z.array(z.string()),
      missing_information: z.array(z.string()),
      uncertainty_flags: z.array(uncertaintyFlagSchema),
      reviewer_questions: z.array(z.string()),
      alternative_candidates: z.array(
        z.object({
          eccn: z.string(),
          title: z.string(),
          status: z.enum(['considered', 'excluded', 'open']),
          rationale: z.string(),
        }),
      ),
      review_path_key: z.string().nullable().optional(),
    }),
  ),
  memo_markdown: z.string(),
  artifacts: z.object({
    extracted_text_path: z.string(),
    structured_output_path: z.string(),
    memo_path: z.string(),
  }),
  run_metadata: z.record(z.string(), z.unknown()).optional().nullable(),
});

export type DocumentCreateInput = z.infer<typeof documentCreateSchema>;
export type ClassificationRunCreateInput = z.infer<typeof classificationRunCreateSchema>;
export type ReviewSubmissionInput = z.infer<typeof reviewSubmissionSchema>;
export type PublicDemoPublishInput = z.infer<typeof publicDemoPublishSchema>;
export type FactReviewUpdateInput = z.infer<typeof factReviewUpdateSchema>;
export type ReviewPathUpdateInput = z.infer<typeof reviewPathUpdateSchema>;
export type ECCNCandidateUpdateInput = z.infer<typeof eccnCandidateUpdateSchema>;
export type WorkerOutput = z.infer<typeof workerOutputSchema>;
export type WorkerCliOutput = z.infer<typeof workerCliOutputSchema>;
export type SignUpInput = z.infer<typeof signUpSchema>;
export type SignInInput = z.infer<typeof signInSchema>;
export type VerifyEmailInput = z.infer<typeof verifyEmailSchema>;
export type ResendVerificationInput = z.infer<typeof resendVerificationSchema>;
export type ForgotPasswordInput = z.infer<typeof forgotPasswordSchema>;
export type ResetPasswordInput = z.infer<typeof resetPasswordSchema>;
export type ChangePasswordInput = z.infer<typeof changePasswordSchema>;
export type OnboardingInput = z.infer<typeof onboardingSchema>;
export type OrganizationUpdateInput = z.infer<typeof organizationUpdateSchema>;
export type InviteCreateInput = z.infer<typeof inviteCreateSchema>;
export type InviteAcceptInput = z.infer<typeof inviteAcceptSchema>;
